python_modules/dagster/dagster/_scheduler/scheduler.py

```diff
@@ -282,46 +286,72 @@ def launch_scheduled_runs(
 
     tick_retention_settings = instance.get_tick_retention_settings(InstigatorType.SCHEDULE)
 
-    schedules: Dict[str, ExternalSchedule] = {}
+    running_schedules: Dict[str, ExternalSchedule] = {}
+    all_workspace_schedule_selector_ids = set()
     error_locations = set()
 
+    now_timestamp = end_datetime_utc.timestamp()
+
     for location_entry in workspace_snapshot.values():
         code_location = location_entry.code_location
         if code_location:
             for repo in code_location.get_repositories().values():
                 for schedule in repo.get_external_schedules():
                     selector_id = schedule.selector_id
+                    all_workspace_schedule_selector_ids.add(selector_id)
                     if schedule.get_current_instigator_state(
                         all_schedule_states.get(selector_id)
                     ).is_running:
-                        schedules[selector_id] = schedule
+                        running_schedules[selector_id] = schedule
+                    elif all_schedule_states.get(selector_id):
+                        schedule_state = all_schedule_states[selector_id]
+                        # If there is a DB row to update, see if we should still update the
+                        # last_iteration_timestamp
+                        _write_and_get_next_checkpoint_timestamp(
+                            instance,
+                            all_schedule_states[selector_id],
+                            cast(ScheduleInstigatorData, schedule_state.instigator_data),
+                            now_timestamp,
+                        )
+
         elif location_entry.load_error:
             error_locations.add(location_entry.origin.location_name)
 
-    # Remove any schedule states that were previously created with DECLARED_IN_CODE
-    # and can no longer be found in the workspace (so that if they are later added
-    # back again, their timestamps will start at the correct place)
+    # Remove any schedule states that were previously created and can no longer
+    # be found in the workspace (so that if they are later added back again,
+    # their timestamps will start at the correct place)
```

Comment:
ok so just to talk it out - one behavior change here is that before, if you had a DECLARED_IN_CODE schedule and turned it off via code, it would *immediately* be removed from the DB (since it had DECLARED_IN_CODE status and was no longer in the "schedules" dict, which should have been called "running_schedules". Now, it will be the same as if you turned any other schedule off, in that it will wait for 12 hours before deleting the row. But because we are now updating the last_iteration_timestamp timestamp, it's less likely to cause problems when it is added back in.

You can construct edge cases where this still doesn't behave as expected right? Like let's say you have a schedule that fires every 5 minutes, you turn it on with default status running, you turn it back to default status STOPPED for 20 minutes, you turn it back to default status RUNNING - it will now try to backfill one tick, and it wouldn't have before, right? That is admittedly far fetched and maybe we don't care enough in practice, but if we were worried about that, we could keep the old 'immediately delete if not running and DECLARED_IN_CODE' behavior for schedules.

Reply:
I'm not sure what the right answer is.

I just think we're exposed to inconsistencies and can construct weird corners right now.

Let's say we have 2 schedules A, B.  Both have declared in code "stopped".  I turn on A manually and turn it off again (note, off, not reset).  Then I remove both and add them both back again with declared in code "started".

The fact that I get different behavior based on schedules A and B (declared in code stopped and manually stopped) just feels a little weird to me.  It relies on the subtle difference of hitting the off button vs hitting the "reset schedule" button.

To be clear, I don't think having these inconsistencies in either direction is strictly a bad idea.  But I can't reason a way to a 100% buttoned-down solution with zero surprises.

Reply:
got it - i think you could consider retaining the old 'immediately delete DECLARED_IN_CODE once it is no longer running' and separately also add waiting 12 hrs before deleting things once they are no longer in the workspace at all, but not a hill I will die on